workflow/controller/operator.go

```diff
@@ -1204,64 +1192,65 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatu
 		}
 	}
 
-	if !node.Completed() {
-		if newDaemonStatus != nil {
-			if !*newDaemonStatus {
-				// if the daemon status switched to false, we prefer to just unset daemoned status field
-				// (as opposed to setting it to false)
-				newDaemonStatus = nil
-			}
-			if (newDaemonStatus != nil && node.Daemoned == nil) || (newDaemonStatus == nil && node.Daemoned != nil) {
-				woc.log.Infof("Setting node %v daemoned: %v -> %v", node.ID, node.Daemoned, newDaemonStatus)
-				node.Daemoned = newDaemonStatus
-				updated = true
-				if pod.Status.PodIP != "" && pod.Status.PodIP != node.PodIP {
-					// only update Pod IP for daemoned nodes to reduce number of updates
-					woc.log.Infof("Updating daemon node %s IP %s -> %s", node.ID, node.PodIP, pod.Status.PodIP)
-					node.PodIP = pod.Status.PodIP
-				}
-			}
-		}
+	// only update Pod IP for daemoned nodes to reduce number of updates
+	if !new.Completed() && new.IsDaemoned() {
+		new.PodIP = pod.Status.PodIP
+	}
 
-		// we only need to update these values if the container transitions to complete
-		if newPhase.Fulfilled() {
-			// outputs are mixed between the annotation (parameters, artifacts, and result) and the pod's status (exit code)
-			if exitCode := getExitCode(pod); exitCode != nil {
-				woc.log.Infof("Updating node %s exit code %d", node.ID, *exitCode)
-				node.Outputs = &wfv1.Outputs{ExitCode: pointer.StringPtr(fmt.Sprintf("%d", int(*exitCode)))}
-				if outputStr, ok := pod.Annotations[common.AnnotationKeyOutputs]; ok {
-					woc.log.Infof("Setting node %v outputs: %s", node.ID, outputStr)
-					if err := json.Unmarshal([]byte(outputStr), node.Outputs); err != nil { // I don't expect an error to ever happen in production
-						node.Phase = wfv1.NodeError
-						node.Message = err.Error()
-					}
-				}
-			}
+	// both exit code and outputs maybe know before the pod terminates, we risk loosing them to pod being deleted
+	// if we do not capture them
+	if exitCode := getExitCode(pod); exitCode != nil {
```

Comment:
I've removed the guard condition on outputs, so we could capture them sooner, reducing impact of "pod deleted"